### PR TITLE
Slider Component Performance Improvements

### PR DESCRIPTION
**Pull request by iethree**
This updates the slider component to improve performance by only calling its `onChange` handler on `mouseUp` or `keyUp` events.  This prevents multiple state changes per second from bubbling out of this component.

To improve UI and give instant feedback while dragging, this also adds onHover tooltips to both sliders.

https://user-images.githubusercontent.com/30528226/172706961-845be2c6-9c5d-4edb-a7e3-964e804be4a0.mp4

You can view this "Slider" component using `yarn storybook`

### CODE REVIEW

File: frontend/src/metabase/core/components/Slider/Slider.tsx [hunk 1]
```diff
@@ -18,76 +22,121 @@ export type NumericInputAttributes = Omit<
 >;
 
 export interface SliderProps extends NumericInputAttributes {
-  value: number[];
-  onChange: (value: number[]) => void;
+  value: (number | undefined)[];
+  onChange: (value: (number | undefined)[]) => void;
```
Comment (iethree) [inline frontend/src/metabase/core/components/Slider/Slider.tsx:26]: allow undefined values to be passed, but internall, we set them to min or max

File: frontend/src/metabase/core/components/Slider/Slider.tsx [hunk 2]
```diff
@@ -18,76 +22,121 @@ export type NumericInputAttributes = Omit<
...
+  const [value, setValue] = useState([
+    parentValue[0] ?? parentMin,
+    parentValue[1] ?? parentMax,
+  ]);
 
-  const [beforeRange, rangeWidth] = useMemo(() => {
-    const totalRange = rangeMax - rangeMin;
+  // potentially expand the input range to include out-of-range values
+  const [min, max] = useMemo(() => {
+    return [_.min([...value, parentMin]), _.max([...value, parentMax])];
```
Comment (iethree) [inline frontend/src/metabase/core/components/Slider/Slider.tsx:47]: the underscore `min` and `max` functions handle `undefined` values better than `Math.min` and `Math.max`

File: frontend/src/metabase/core/components/Slider/Slider.tsx [hunk 3]
```diff
@@ -18,76 +22,121 @@ export type NumericInputAttributes = Omit<
...
-  }, [value, rangeMin, rangeMax]);
+  }, [value, min, max]);
 
-  const handleChange = useCallback(
+  const handleInput = useCallback(
     (event: ChangeEvent<HTMLInputElement>, valueIndex: number) => {
       const changedValue = [...value];
       changedValue[valueIndex] = Number(event.target.value);
-      onChange(changedValue);
+      setValue(changedValue);
```
Comment (iethree) [inline frontend/src/metabase/core/components/Slider/Slider.tsx:66]: now, on change we simply save the values in this component's local state

File: frontend/src/metabase/core/components/Slider/Slider.tsx [hunk 4]
```diff
@@ -18,76 +22,121 @@ export type NumericInputAttributes = Omit<
...
 
-  const sortValues = useCallback(() => {
-    if (value[0] > value[1]) {
-      const sortedValues = [...value].sort();
-      onChange(sortedValues);
-    }
+  const handleChange = useCallback(() => {
+    const sortedValues = value[1] < value[0] ? [...value].sort() : value;
+    onChange(sortedValues);
   }, [value, onChange]);
```
Comment (iethree) [inline frontend/src/metabase/core/components/Slider/Slider.tsx:74]: we only pass this component's values up to the parent component on `mouseUp` or `keyUp` to keep the number of updates reasonable.

File: frontend/src/metabase/core/components/Slider/Slider.tsx [hunk 5]
```diff
@@ -18,76 +22,121 @@ export type NumericInputAttributes = Omit<
...
+        min={min}
+        max={max}
         step={step}
       />
     </SliderContainer>
   );
 };
 
+const getTooltipPosition = (basePosition: number) =>
+  `calc(${basePosition}% + ${11 - basePosition * 0.18}px)`;
```
Comment (iethree) [inline frontend/src/metabase/core/components/Slider/Slider.tsx:138]: This is a gnarly bit of math to line the tooltip up with the center of the drag handle

File: frontend/src/metabase/core/components/Slider/Slider.tsx (hunk 2 above)
Comment (npfitz) [reply frontend/src/metabase/core/components/Slider/Slider.tsx:47]: Haha, it's actually surprising the difference between the two (just gave it a try). Would it just be a matter of filtering out undefined from `value` though to use `Math.min`?

Comment (iethree) [reply frontend/src/metabase/core/components/Slider/Slider.tsx:47]: yeah, i actually did that in a prior version, but this is much easier to read.

File: frontend/src/metabase/core/components/Slider/Slider.tsx (hunk 5 above)
Comment (WiNloSt) [reply frontend/src/metabase/core/components/Slider/Slider.tsx:138]: #### I don't think it's worth changing considering how your simple calculation is almost correct like 99.9%

Nit: I believe this math is somewhat off. Like it's not totally off, but just by a bit. Basically, the width of the tooltip container is larger than the length the slider thumb can travel by the thumb width.

That means we need to scale the track 100% down to the thumb travel length. And that's not possible via CSS. From my research, we can't put any percentage as a divisor.

So what I came up with is to put tooltips in a new container and make the container the size of the slider thumb travel length. Then we can easily match the position of the tooltip with the slider thumb.
https://www.loom.com/share/33ce4de78604438d81343b2dde0cc04f

File: frontend/src/metabase/core/components/Slider/Slider.tsx [hunk 6]
```diff
@@ -18,76 +22,121 @@ export type NumericInputAttributes = Omit<
...
-    if (value[0] > value[1]) {
-      const sortedValues = [...value].sort();
-      onChange(sortedValues);
-    }
+  const handleChange = useCallback(() => {
+    const sortedValues = value[1] < value[0] ? [...value].sort() : value;
+    onChange(sortedValues);
   }, [value, onChange]);
 
+  const [minValue, maxValue] = useMemo(
```
Comment (WiNloSt) [inline frontend/src/metabase/core/components/Slider/Slider.tsx:76]: I find it a little hard to keep track of all the slightly different variables while trying to reason about the code. e.g. `min`, `parentMin`, `value` (which uses in `Math.min(...value)`), `minValue`. Not sure what we can do here, but it's just my observation.

**Review by WiNloSt — COMMENTED** [review]
The Slider looks awesome!

File: frontend/src/metabase/core/components/Slider/Slider.stories.tsx [hunk 7]
```diff
@@ -8,8 +8,12 @@ export default {
 };
 
 const Template: ComponentStory<typeof Slider> = args => {
-  const [value, setValue] = useState([10, 40]);
-  return <Slider {...args} value={value} onChange={setValue} />;
+  const [value, setValue] = useState<(number | undefined)[]>([10, 40]);
+  return (
+    <div className="pt4">
+      <Slider {...args} value={value} onChange={setValue} />
```
[the remark this answers is not in the document]
Comment (iethree) [reply frontend/src/metabase/core/components/Slider/Slider.stories.tsx:14]: cool! I didn't know about storybook actions

File: frontend/src/metabase/core/components/Slider/Slider.tsx [hunk 8]
```diff
@@ -18,76 +22,121 @@ export type NumericInputAttributes = Omit<
...
       <SliderInput
         type="range"
         aria-label="min"
         value={value[0]}
-        onChange={e => handleChange(e, 0)}
-        onMouseUp={sortValues}
-        min={rangeMin}
-        max={rangeMax}
+        onChange={e => handleInput(e, 0)}
+        onMouseUp={handleChange}
```
[the remark this answers is not in the document]
Comment (iethree) [reply frontend/src/metabase/core/components/Slider/Slider.tsx:107]: good catch, adding `onTouchEnd` as well

File: frontend/src/metabase/core/components/Slider/Slider.tsx (hunk 6 above)
Comment (iethree) [reply frontend/src/metabase/core/components/Slider/Slider.tsx:76]: 🤔  part of this is the inherent complexity of juggling different internal + external values, but I think I can make some improvements here too 👍

File: frontend/src/metabase/core/components/Slider/Slider.tsx (hunk 5 above)
Comment (iethree) [reply frontend/src/metabase/core/components/Slider/Slider.tsx]: your way seems better, and with styled components, I don't think it actually makes the code harder to read.

You also got me looking at things WAY zoomed in so I made a couple other geometry tweaks. html range inputs are weird.

File: frontend/src/metabase/core/components/Slider/Slider.unit.spec.tsx [hunk 9]
```diff
@@ -23,21 +23,42 @@ describe("Slider", () => {
...
 
     const [firstCall, secondCall] = spy.mock.calls;
 
     expect(firstCall[0]).toEqual([5, 20]);
-    expect(secondCall[0]).toEqual([10, 15]);
+    expect(secondCall[0]).toEqual([5, 15]);
+  });
+
+  // handling input via arrow keys
+  it("should call onChange with the new value on keyup", () => {
```
Comment (WiNloSt) [inline frontend/src/metabase/core/components/Slider/Slider.unit.spec.tsx:46]: Oh cool. Just saw this component also supports keyboard actions which are great for accessibility but having some visual cues to indicate which slider (thumb) is active would be even better.

**Review by WiNloSt — APPROVED** [review]

**Pull request merged**